ibc/host_introspection_test.go

```diff
@@ -0,0 +1,271 @@
+		},
+		{
+			name: "invalid: different proof spec",
+			pcs: &light_client_types.PocketClientState{
+				NetworkId:       block.BlockHeader.NetworkId,
+				TrustLevel:      &light_client_types.Fraction{Numerator: 2, Denominator: 3},
+				TrustingPeriod:  durationpb.New(unbondingPeriod),
+				UnbondingPeriod: durationpb.New(unbondingPeriod),
+				MaxClockDrift:   durationpb.New(approxTime),
+				LatestHeight: &client_types.Height{
+					RevisionNumber: 1,
+					RevisionHeight: 0,
+				},
+				ProofSpec: types.ConvertFromIcs23ProofSpec(ics23.IavlSpec),
+			},
+			expectedErr: errors.New("counterparty client state has different proof spec"),
+		},
+		{
+			name: "invalid: different unbonding period",
+			pcs: &light_client_types.PocketClientState{
+				NetworkId:       block.BlockHeader.NetworkId,
+				TrustLevel:      &light_client_types.Fraction{Numerator: 2, Denominator: 3},
+				TrustingPeriod:  durationpb.New(unbondingPeriod),
+				UnbondingPeriod: durationpb.New(unbondingPeriod + 1),
+				MaxClockDrift:   durationpb.New(approxTime),
+				LatestHeight: &client_types.Height{
+					RevisionNumber: 1,
+					RevisionHeight: 0,
+				},
+				ProofSpec: types.SmtSpec,
+			},
+			expectedErr: errors.New("counterparty client state has different unbonding period"),
+		},
+		{
+			name: "invalid: unbonding period less than trusting period",
+			pcs: &light_client_types.PocketClientState{
+				NetworkId:       block.BlockHeader.NetworkId,
+				TrustLevel:      &light_client_types.Fraction{Numerator: 2, Denominator: 3},
+				TrustingPeriod:  durationpb.New(unbondingPeriod),
+				UnbondingPeriod: durationpb.New(unbondingPeriod - 1),
+				MaxClockDrift:   durationpb.New(approxTime),
+				LatestHeight: &client_types.Height{
+					RevisionNumber: 1,
+					RevisionHeight: 0,
+				},
+				ProofSpec: types.SmtSpec,
+			},
+			expectedErr: errors.New("counterparty client state unbonding period is less than trusting period"),
+		},
+		{
+			name: "valid client state",
+			pcs: &light_client_types.PocketClientState{
+				NetworkId:       block.BlockHeader.NetworkId,
+				TrustLevel:      &light_client_types.Fraction{Numerator: 2, Denominator: 3},
+				TrustingPeriod:  durationpb.New(unbondingPeriod),
+				UnbondingPeriod: durationpb.New(unbondingPeriod),
+				MaxClockDrift:   durationpb.New(approxTime),
+				LatestHeight: &client_types.Height{
+					RevisionNumber: 1,
+					RevisionHeight: 0,
+				},
+				ProofSpec: types.SmtSpec,
+			},
+			expectedErr: nil,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			bz, err := codec.GetCodec().Marshal(tc.pcs)
+			require.NoError(t, err)
+			clientState := &client_types.ClientState{
+				Data: bz,
+				RecentHeight: &client_types.Height{
+					RevisionNumber: 1,
+					RevisionHeight: 0,
+				},
+			}
+			err = cm.VerifyHostClientState(clientState)
+			require.ErrorAs(t, err, &tc.expectedErr)
+		})
+	}
+}
```

Comment:
Did not review the tests in detail but I looked through all the names and it's great coverage 👌